File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:  
      stack<pair<int, int>> s;
	int mini;
    MinStack() {
  mini = INT_MAX;
    }
    
    void push(int x) {
        mini = min(x, mini);
		s.push({x, mini});
    }
    
    void pop() {
        	s.pop();
		if (s.empty())
			mini = INT_MAX;
		else
			mini = s.top().second;
    }
    
    int top() {
        	 if (s.empty())
            return -1;
        return s.top().first ;
    }
    
    int getMin() {
      	 if (s.empty())
            return -1;
        return s.top().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(value);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.py
from collections import deque

class MinStack:

    def __init__(self):
        self.dq = deque()
        self.mini = float("inf")

    def push(self, value: int) -> None:
        self.mini = min(value, self.mini)
        self.dq.append((value, self.mini))

    def pop(self) -> None:
        if not self.dq:
            return

        self.dq.pop()

        if not self.dq:
            self.mini = float("inf")
        else:
            self.mini = self.dq[-1][1]

    def top(self) -> int:
        if not self.dq:
            return -1

        return self.dq[-1][0]

    def getMin(self) -> int:
        if not self.dq:
            return -1

        return self.dq[-1][1]